### ISSUE TITLE
Suggest DeriveAnyClass

### ISSUE THREAD

**Issue by ocharles** [ndmitchell/hlint#956]
We can write

```haskell
data Foo = { x, y, z :: T }

instance C1 Foo

instance C2 Foo
```

as just

```haskell
data Foo = { x, y, z :: T } deriving (C1, C2)
```

Could HLint suggest this?

**Comment by ndmitchell** [comment]
I'm wary about doing so because the rules around DeriveAnyClass vs GeneralizedNewtypeDeriving have changed in recent history and the difference between `instance C1 Foo` and `instance C1 Foo where` is super subtle. The GHC semantics here are way too much like "implicit magic that can easily bite you in the ass", and when you take that and suggest transformations, I'm really worried about screwing up the code in subtle ways.

**Comment by ocharles** [comment]
Is it more subtle than always suggesting `deriving anytype` rather than just `deriving`?

**Comment by ndmitchell** [comment]
Suggesting `deriving anytype` would be a lot more obvious. But `anytype` is quite new, so not available as far back as the versions HLint would like to suggest valid Haskell for. I think adding DeriveAnyClass without making it require an explicit anytype was a bad idea, and using a derive any class without it scares me, but it will be a while before the old versions have gone and deriving strategies is universally adopted.

**Comment by ocharles** [comment]
This is reasonable caution! But HLint can have stuff off by default, right - and people could opt in to this if they wanted? Suggesting `deriving anytype` would avoid introducing any broken code, providing that the code originally already used deriving strategies (which you could reach by enabling the 8.10 warning to warn on un-annotated deriving strategies).

**Comment by ocharles** [comment]
Another option I'd be happy with is to only suggest this for types that can be found to be `data`, not `newtype`, which I think is much less risky. But maybe less tractable for HLint.

**Comment by googleson78** [comment]
Sorry for derailing slightly but could I ask what the aforementioned difference is (no `where` vs having `where`)? I couldn't find it documented in `ghc` or by testing my first guess manually ;/

FWIW I agree with "`anytype` derivations should always have an explicit strategy"

**Comment by ndmitchell** [comment]
@googleson78 I'm not sure. I think one is equivalent to not implementing any methods, and one to using generic instances? Maybe there isn't a difference anymore? It's certainly bit me in the past.

@ocharles figuring out data/newtype is easy in this instance - it had better be in the same module, and you'd better be able to find the definition, otherwise you can't make the suggestion. Usually it's hard for HLint, but this case is different. Suggesting it but off by default is fine - patch welcome - but those changes tend not to have much impact outside specific code bases.

**Comment by ocharles** [comment]
Argh, I just realised it's even more complicated than just using `derive anytype`. That needs `-XDeriveAnyType` enabled, but that then _overrides_ GND for any _other_ `newtype` instance declarations. I'm not coming round to agreeing that maybe it's best we shelve this (close the issue). I'm unlikely to have the time to get into this as it's much harder than I thought (despite you clearly telling me!)

**Comment by googleson78** [comment]
@ndmitchell 
For `aeson`s `ToJSON`, they seem to be the same:
<details><summary>Example</summary>

```haskell
blademaster :: ~ » cat aeson.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE DerivingStrategies #-}

import Data.Aeson
import GHC.Generics (Generic)

data Foo0 =
  Foo0
    { bar0 :: Int
    }
  deriving stock Generic

instance ToJSON Foo0

data Foo1 =
  Foo1
    { bar1 :: Int
    }
  deriving stock Generic

instance ToJSON Foo1 where

data Foo2 =
  Foo2
    { bar2 :: Int
    }
  deriving stock Generic
  deriving anyclass ToJSON
blademaster :: ~ » ghci aeson.hs
GHCi, version 8.6.5: http://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /home/googleson78/.ghci
[1 of 1] Compiling Main             ( aeson.hs, interpreted )
Ok, one module loaded.
> toJSON $ Foo0 42
Object (fromList [("bar0",Number 42.0)])
> toJSON $ Foo1 42
Object (fromList [("bar1",Number 42.0)])
> toJSON $ Foo2 42
Object (fromList [("bar2",Number 42.0)])
```
</details>

But I definitely agree that all the rules surrounding which deriving clause to use and such are too finicky to be left without an explicit strategy

**Comment by ocharles** [comment]
You need to use `newtype` to witness the difference. `anyclass ToJSON` for `newtype Foo2` will be different from `newtype ToJSON Foo2`.

**Comment by googleson78** [comment]
I only wanted to show the difference (or lack thereof?) between `instance X` and `instance X where`. I shouldn't have included the derived version I guess.

**Comment by ndmitchell** [comment]
I have no idea what the difference is, all I know is something hurt me in the past, and I think it was missing a where, but the details are hazy and the scars from instance declarations are real.
